### DISCUSSION TITLE
Largest Neural Network that can be synthesized?

### DISCUSSION THREAD

**Discussion by satvikmaurya** [fastmachinelearning/hls4ml#665] [Q&A]
Hi, 

I'm trying to synthesize a neural network with 3 hidden layers for my current project but I'm getting the following error:

```
**ERROR: [XFORM 203-103] Array 'mult.V' (firmware/nnet_utils/nnet_dense_latency.h:39): partitioned elements number (10000) has exeeded the threshold (4096), which may cause long run-time.**
ERROR: [HLS 200-70] Pre-synthesis failed.
command 'ap_source' returned error code
    while executing
"source build_prj.tcl"
    ("uplevel" body line 1)
    invoked from within
"uplevel \#0 [list source $arg] "
```

The neural network I'm trying to synthesize is this:

```
input_size = 100

model = Sequential()
model.add(QDense(input_size, input_shape=(input_size,), name='fc1',
                 kernel_quantizer=quantized_bits(6,0,alpha=1), bias_quantizer=quantized_bits(6,0,alpha=1),
                 kernel_initializer='lecun_uniform', kernel_regularizer=l1(0.0001)))
model.add(QActivation(activation=quantized_relu(6), name='relu1'))
model.add(QDense(input_size/2, name='fc2',
                 kernel_quantizer=quantized_bits(6,0,alpha=1), bias_quantizer=quantized_bits(6,0,alpha=1),
                 kernel_initializer='lecun_uniform', kernel_regularizer=l1(0.0001)))
model.add(QActivation(activation=quantized_relu(6), name='relu2'))
model.add(QDense(input_size/2, name='fc3',
                 kernel_quantizer=quantized_bits(6,0,alpha=1), bias_quantizer=quantized_bits(6,0,alpha=1),
                 kernel_initializer='lecun_uniform', kernel_regularizer=l1(0.0001)))
model.add(QActivation(activation=quantized_relu(6), name='relu3'))
model.add(QDense(32, name='output',
                 kernel_quantizer=quantized_bits(6,0,alpha=1), bias_quantizer=quantized_bits(6,0,alpha=1),
                 kernel_initializer='lecun_uniform', kernel_regularizer=l1(0.0001)))
model.add(Activation(activation='softmax', name='softmax'))
```

Is there a limit on the size (breadth of the network in this case) that can be synthesized? I am using Vivado HLS 2019.1.

**Comment by vloncar** [comment]
Limit depends on many factors, such as network size (number of parameters), network architecture, quantization, parallelization, target FPGA... So it is not easy to say. In your case, I would switch to the resource strategy to avoid the 4096 limit. With `hls_config['Model']['Strategy'] = 'Resource'`. Also, increase the reuse factor, RF=1 will not work.

**Reply by satvikmaurya** [reply]
Thanks!

I tried the setting above, and the error I'm getting now is this: 

`ERROR: [XFORM 203-504] Stop unrolling loop 'MultLoop' (firmware/nnet_utils/nnet_dense_resource.h:73) in function 'nnet::dense_resource<ap_fixed<16, 6, (ap_q_mode)0, (ap_o_mode)0, 0>, ap_fixed<16, 6, (ap_q_mode)5, (ap_o_mode)3, 0>, config8>' because it may cause large runtime and excessive memory usage due to increase in code size. Please avoid unrolling the loop or form sub-functions for code in the loop body. `

Is there anything else I can try? I increased the RF to 25 too.

**Reply by vloncar** [reply]
You should try with a small network first and if that fits then try with progressively larger networks.
